src/nodelets/obstacles_detection.cpp

```diff
@@ -168,14 +159,144 @@ class ObstaclesDetection : public nodelet::Nodelet
+		}
+
+		//Common variables for all strategies
+		pcl::PointCloud<pcl::PointXYZ>::Ptr hypotheticalGroundCloud(new pcl::PointCloud<pcl::PointXYZ>);
+		pcl::PointCloud<pcl::PointXYZ>::Ptr obstaclesCloud(new pcl::PointCloud<pcl::PointXYZ>);
+		pcl::IndicesPtr ground, obstacles;
+		pcl::PointCloud<pcl::PointXYZ>::Ptr groundCloud(new pcl::PointCloud<pcl::PointXYZ>);
+
+		ros::Time lasttime = ros::Time::now();
+
+		if (!simpleSegmentation_ && !optimizeForCloseObject_){
+			//This is the default strategy
+			//The cloud is divided into two pointcloud based on reported Z and the position of the camera.
+			// One is the hypothetical ground cloud and the other one is the obstacles pointcloud.
+			//The algorithm then extracts (and remove) from the hypothetical ground cloud the detected obstacles,
+			//and add them to the obstacles pointcloud
+			originalCloud = rtabmap::util3d::transformPointCloud(originalCloud, localTransform);
+			hypotheticalGroundCloud = rtabmap::util3d::passThrough(originalCloud, "z", std::numeric_limits<int>::min(), maxFloorHeight_);
+
+			rtabmap::util3d::segmentObstaclesFromGround<pcl::PointXYZ>(hypotheticalGroundCloud,
+					ground, obstacles, normalEstimationRadius_, groundNormalAngle_, minClusterSize_);
+
+			if(ground.get() && ground->size())
+			{
+				pcl::copyPointCloud(*hypotheticalGroundCloud, *ground, *groundCloud);
+			}
+
+			obstaclesCloud = rtabmap::util3d::passThrough(originalCloud, "z", maxFloorHeight_, maxObstaclesHeight_);
+
+			if(obstacles.get() && obstacles->size())
+			{
+				pcl::PointCloud<pcl::PointXYZ>::Ptr obstaclesNearFloorCloud(new pcl::PointCloud<pcl::PointXYZ>);
+				pcl::copyPointCloud(*hypotheticalGroundCloud, *obstacles, *obstaclesNearFloorCloud);
+				*obstaclesCloud += *obstaclesNearFloorCloud;
+			}
+
+		}
+		if (!simpleSegmentation_ && optimizeForCloseObject_){
+			// If the option optimize for close object has been set to true,
+			// we divide the floor point cloud into two subsections, one for all potential floor points up to 1m
+			// one for potential floor points further away than 1m.
+			// For the points at closer range, we use a smaller normal estimation radius and ground normal angle,
+			// which allows to detect smaller objects, without increasing the number of false positive.
+			// For all other points, we use a biger normal estimation radius (* 3.) and a bigger tolerance for the
+			// grond normal angle (* 2.).
+
+			originalCloud = rtabmap::util3d::transformPointCloud(originalCloud, localTransform);
+			pcl::PointCloud<pcl::PointXYZ>::Ptr originalCloud_front = rtabmap::util3d::passThrough(originalCloud, "x", std::numeric_limits<int>::min(), 1.);
+			pcl::PointCloud<pcl::PointXYZ>::Ptr originalCloud_back = rtabmap::util3d::passThrough(originalCloud, "x", 1., std::numeric_limits<int>::max());
+
+			pcl::PointCloud<pcl::PointXYZ>::Ptr hypotheticalGroundCloud_front = rtabmap::util3d::passThrough(originalCloud_front, "z", std::numeric_limits<int>::min(), maxFloorHeight_);
```

Comment:
Filter by z first, then separate front and back, saves time and simplifies the code